File: scripts/sistema/login/login_test.py
# Importar las librerías necesarias 
import time
## Importar unittest para las pruebas
import unittest
## Importar Selenium y las herramientas necesarias para manejar Chrome
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager

# Defino una clase que contendrá todas las pruebas relacionadas con el login
class LoginTest(unittest.TestCase):

    # Cada función será una prueba diferente
    def test_login_con_datos_correctos(self):
        # Instalación del chromedriver
        service = Service(ChromeDriverManager().install())

        # Arrancar el navegador Chrome
        driver = webdriver.Chrome(service=service)

        # Voy a navegar a la página de login
        driver.get("https://katalon-demo-cura.herokuapp.com/profile.php#login")

        # Rellenar el campo de usuario
        driver.find_element("id", "txt-username").send_keys("John Doe")
        # Rellenar el campo de contraseña
        driver.find_element("id", "txt-password").send_keys("ThisIsNotAPassword")
        # Hacer click en el botón de login
        driver.find_element("xpath", "//*[@id='btn-login']").click()

        # Vamos a esperar unos segundos
        # driver.implicitly_wait(5)  # Espera implícita de 5 segundos
        # Esto no está funcionando y es un error típico en Selenium... la espera
        # Vamos a forzar esa espera desde python
        time.sleep(5)  # Espera explícita de 5 segundos

        # Vamos a ver a qué página hemos llegado
        urlNuevaPagina = driver.current_url

        # Comprobar que la URL es la esperada tras un login exitoso
        self.assertEqual("https://katalon-demo-cura.herokuapp.com/#appointment", urlNuevaPagina)

        # Quiero asegurarme que no aparece ningún mensaje de error
        # "Login failed! Please ensure the username and password are valid."... XPATH (párrafo que contenga el texto)
        mensajesError = driver.find_elements("xpath", "//p[contains(text(), 'Login failed! Please ensure the username and password are valid.')]")
        self.assertEqual(0, len(mensajesError))  # No debería haber mensajes de error

        # Cierra el navegador
        driver.quit()



if __name__ == '__main__':
    unittest.main() 

File: scripts/sistema_js/tests/login_test.js
const fs = require('fs');
const path = require('path');
const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const { expect } = require('chai');

const { describe, it, beforeEach, afterEach } = require('mocha');
// Lo que hace es instalar chromedriver
require('chromedriver');


describe('Login Test Suite', function() {

    const USUARIO_CORRECTO = "John Doe"
    const CONTRASENA_CORRECTA = "ThisIsNotAPassword"
    const URL_LOGIN = "https://katalon-demo-cura.herokuapp.com/profile.php#login"
    const URL_LOGIN_NO_EXITOSO = "https://katalon-demo-cura.herokuapp.com/profile.php#login"
    const URL_LOGIN_EXITOSO = "https://katalon-demo-cura.herokuapp.com/#appointment"
    const MENSAJE_ERROR_LOGIN = "Login failed! Please ensure the username and password are valid."

    const ID_CAMPO_USUARIO = "txt-username"
    const ID_CAMPO_CONTRASENA = "txt-password"
    const XPATH_BOTON_LOGIN = "//*[@id='btn-login']"
    const XPATH_MENSAJE_ERROR = "//p[contains(text(), '" + MENSAJE_ERROR_LOGIN + "')]"

    const TIEMPO_ESPERA = 5

    const ANCHO_PANTALLA = 1024
    const ALTO_PANTALLA = 768

    const CARPETA_CAPTURAS = "./capturas"
    let driver;

    beforeEach(async function() {
        // Crear carpeta para capturas si no existe
        if (!fs.existsSync(CARPETA_CAPTURAS)){
            fs.mkdirSync(CARPETA_CAPTURAS);
        }

        // Configurar el driver de Chrome
        let options = new chrome.Options();
        options.addArguments(`--window-size=${ANCHO_PANTALLA},${ALTO_PANTALLA}`);
        options.addArguments('--headless'); // Ejecutar en modo headless si es necesario

        driver = await new Builder()
            .forBrowser('chrome')
            .setChromeOptions(options)
            .build();

        await driver.get(URL_LOGIN);
        // Ancho y alto
        await driver.manage().window().setRect({ width: ANCHO_PANTALLA, height: ALTO_PANTALLA });
    });

    afterEach(async function() {
        // Captura de pantalla en caso de fallo
        const nombreCaptura = path.join(CARPETA_CAPTURAS, `despues_login_${Date.now()}.png`);
        const imagen = await driver.takeScreenshot();
        fs.writeFileSync(nombreCaptura, imagen, 'base64');

        await driver.quit();
    });

    it('Login exitoso con credenciales correctas', async function() {
        // CUANDO
        // Meto en el formulario de login el usuario en el input cuyo id es "usuario"
        await driver.findElement(By.id(ID_CAMPO_USUARIO)).sendKeys(USUARIO_CORRECTO);
        // Meto en el formulario de login la contraseña en el input cuyo id es "password"
        await driver.findElement(By.id(ID_CAMPO_CONTRASENA)).sendKeys(CONTRASENA_CORRECTA);
        // Hacer click en el botón de login
        await driver.findElement(By.xpath(XPATH_BOTON_LOGIN)).click();

        // Captura de pantalla tras el login
        const nombreCaptura = path.join(CARPETA_CAPTURAS, `antes_login_${Date.now()}.png`);
        const imagen = await driver.takeScreenshot();
        fs.writeFileSync(nombreCaptura, imagen, 'base64');

        // ENTONCES
        // Quiero asegurarme que no aparece ningún mensaje de error
        // "Login failed! Please ensure the username and password are valid."... XPATH (párrafo que contenga el texto)
        const mensajesError = await driver.findElements(By.xpath(XPATH_MENSAJE_ERROR));
        expect(mensajesError.length).to.equal(0);  // No debería haber mensajes de error
        // Tengo que llegar a una página principal: https://katalon-demo-cura.herokuapp.com/#appointment
        const urlActual = await driver.getCurrentUrl();
        expect(urlActual).to.equal(URL_LOGIN_EXITOSO);
    });
});